pyanaconda/modules/subscription/runtime.py

```diff
@@ -580,3 +602,327 @@ def run(self):
+                # refresh the ProxyString internal URL cache after setting the port number
+                proxy.parse_components()
+            proxy_url = str(proxy)
+        else:
+            proxy_url = None
+
+        # create the download task
+        download_task = DownloadSatelliteProvisioningScriptTask(
+            satellite_url=self._subscription_request.server_hostname,
+            proxy_url=proxy_url
+        )
+
+        # run it
+        try:
+            log.debug("registration attempt: downloading Satellite provisioning script")
+            download_task.run()
+            log.debug("registration attempt: downloaded Satellite provisioning script")
+        except SatelliteProvisioningError as e:
+            log.debug("registration attempt: failed to download Satellite provisioning script")
+            # Failing to download the Satellite provisioning script for a user provided
+            # server hostname is an unrecoverable error (wrong URL or incorrectly configured
+            # Satellite instance), so we end there.
+            raise e
+
+        # before running the Satellite provisioning script we back up the current RHSM config
+        # file state, so that we can restore it if Satellite provisioning rollback become necessary
+        rhsm_config_proxy = self._rhsm_observer.get_proxy(RHSM_CONFIG)
+        backup_task = BackupRHSMConfBeforeSatelliteProvisioningTask(
+            rhsm_config_proxy=rhsm_config_proxy
+        )
+        self._config_backup_callback(backup_task.run())
+
+        # now run the Satellite provisioning script we just downloaded, so that the installation
+        # environment can talk to the Satellite instance the user has specified via custom
+        # server hostname
+        run_script_task = RunSatelliteProvisioningScriptTask()
+        run_script_task.succeeded_signal.connect(
+            lambda: self._registered_to_satellite_callback(True)
+        )
+        try:
+            log.debug("registration attempt: running Satellite provisioning script")
+            run_script_task.run_with_signals()
+            log.debug("registration attempt: Satellite provisioning script has been run")
+            # unfortunately the RHSM service apparently does not pick up the changes done
+            # by the provisioning script to rhsm.conf, so we need to restart the RHSM systemd
+            # service, which will make it re-read the config file
+            # FIXME: find out if we can use something less radical than a full
+            #        systemd service restart
+            util.restart_service(RHSM_SERVICE_NAME)
+
+        except SatelliteProvisioningError as e:
+            log.debug("registration attempt: Satellite provisioning script run failed")
+            # Failing to run the Satellite provisioning script successfully,
+            # which is an unrecoverable error, so we end there.
+            raise e
+
+    def run(self):
```

Comment:
The same works here. I see that you can split `register_task` work from the rest. Maybe even better to have `_create_register_task` and `_run_register_task`.

Reply:
What you see here is a transitional state of the registration task due to payload modularization not being fully complete (eq. other bits and pieces still have to reside elsewhere) and I would like to avoid further refactoring of the register tasks run method until after payload modularization is done.

Then all the things needed for registration can finally reside in the task and can be better re-factored.

Reply:
Ok, thanks for explanation.